File: runinator-command-center/src/utils/__tests__/wdl-expression.test.ts
import { describe, expect, it } from "vitest";
import { expressionJsonToWdl, parseWdlExpression } from "../wdl-expression";

describe("WDL expression conversion", () => {
  it("renders lowered references and operators as WDL surface expressions", () => {
    expect(expressionJsonToWdl({ "$ref": { input: ["ticket_id"] } })).toBe("input.ticket_id");
    expect(expressionJsonToWdl({ "$ref": { workflow: ["attempt"] } })).toBe("run.attempt");
    expect(expressionJsonToWdl({ "$ref": { node: "create_ticket", output: ["id"] } })).toBe("create_ticket.id");
    expect(expressionJsonToWdl({ "$concat": ["ticket ", { "$ref": { input: ["ticket_id"] } }] })).toBe("\"ticket \" ++ input.ticket_id");
    expect(expressionJsonToWdl({ "$coalesce": [{ "$ref": { prev: ["name"] } }, "unknown"] })).toBe("prev.name ?? \"unknown\"");
    expect(expressionJsonToWdl({ "$to_string": { "$ref": { prev: ["count"] } } })).toBe("string(prev.count)");
  });

  it("parses WDL surface expressions back into lowered JSON values", () => {
    expect(parseWdlExpression("input.ticket_id")).toEqual({ "$ref": { input: ["ticket_id"] } });
    expect(parseWdlExpression("\"ticket \" ++ input.ticket_id")).toEqual({ "$concat": ["ticket ", { "$ref": { input: ["ticket_id"] } }] });
    expect(parseWdlExpression("prev.name ?? \"unknown\"")).toEqual({ "$coalesce": [{ "$ref": { prev: ["name"] } }, "unknown"] });
    expect(parseWdlExpression("string(prev.count)")).toEqual({ "$to_string": { "$ref": { prev: ["count"] } } });
    expect(parseWdlExpression("{ message: string(prev.count), tags: [input.tag] }")).toEqual({
      message: { "$to_string": { "$ref": { prev: ["count"] } } },
      tags: [{ "$ref": { input: ["tag"] } }]
    });
  });
});

File: runinator-command-center-flutter/test/core/utils/wdl_expression_test.dart
import 'package:runinator_command_center_flutter/core/utils/wdl_expression.dart';
import 'package:test/test.dart';

void main() {
  group('WDL expression conversion', () {
    test('renders lowered references and operators as WDL surface expressions', () {
      expect(expressionJsonToWdl({r'$ref': {'params': ['ticket_id']}}), 'params.ticket_id');
      expect(expressionJsonToWdl({r'$ref': {'input': ['ticket_id']}}), '{ input: ["ticket_id"] }');
      expect(expressionJsonToWdl({r'$ref': {'workflow': ['attempt']}}), 'run.attempt');
      expect(
        expressionJsonToWdl({r'$ref': {'node': 'create_ticket', 'output': ['id']}}),
        'create_ticket.id',
      );
      expect(
        expressionJsonToWdl({r'$concat': ['ticket ', {r'$ref': {'params': ['ticket_id']}}]}),
        '"ticket " ++ params.ticket_id',
      );
      expect(
        expressionJsonToWdl({r'$coalesce': [{r'$ref': {'prev': ['name']}}, 'unknown']}),
        'prev.name ?? "unknown"',
      );
      expect(
        expressionJsonToWdl({r'$to_string': {r'$ref': {'prev': ['count']}}}),
        'string(prev.count)',
      );
    });

    test('parses WDL surface expressions back into lowered JSON values', () {
      expect(parseWdlExpression('params.ticket_id'), {r'$ref': {'params': ['ticket_id']}});
      expect(parseWdlExpression('"ticket " ++ params.ticket_id'), {
        r'$concat': ['ticket ', {r'$ref': {'params': ['ticket_id']}}],
      });
      expect(parseWdlExpression('input.ticket_id'), {
        r'$ref': {'node': 'input', 'output': ['ticket_id']},
      });
      expect(parseWdlExpression('prev.name ?? "unknown"'), {
        r'$coalesce': [{r'$ref': {'prev': ['name']}}, 'unknown'],
      });
      expect(parseWdlExpression('string(prev.count)'), {
        r'$to_string': {r'$ref': {'prev': ['count']}},
      });
      expect(parseWdlExpression('{ message: string(prev.count), tags: [params.tag] }'), {
        'message': {r'$to_string': {r'$ref': {'prev': ['count']}}},
        'tags': [{r'$ref': {'params': ['tag']}}],
      });
    });
  });
}
